Rando: Even more randomizer settings menu changes

aMannus opened the pull request:
What this does:

- Big changes in the layout of the tabs and columns. See screenshots below.
- Wrap every column inside an ImGui child, stretching every column to the bottom of the window and allowing individual scrollbars on each column.
- Code style changes like making InsertHoverText indentation consistent.
- Change "First use" size for Randomizer Settings menu.
- Change column names under "Tricks/Glitches" tab to "Disabled" and "Enabled" in preparation of new functionality. I toyed with the idea of calling it "Not Considered in Logic" and "Considered in Logic", but it might honestly be clear enough with just "Disabled" and "Enabled" as well. Open for discussion about this one.
- Implement the new incompatible checkbox features for `Skip Child Stealth`, `Shuffle Weird Egg`, `Shuffle Ocarinas` and `Shuffle Kokiri Sword`

Old:
![image](https://user-images.githubusercontent.com/70942617/187920628-6f2186b4-aae7-48d9-824b-db0e4dd4e4d9.png)
![image](https://user-images.githubusercontent.com/70942617/187920693-4771d916-b6d6-49e8-9197-9c99646b42d0.png)
![image](https://user-images.githubusercontent.com/70942617/187920741-0c561339-c8e4-4b44-a1dc-f5b190f5ecbc.png)
![image](https://user-images.githubusercontent.com/70942617/187920795-f7004386-9820-4166-8b17-06847e88fb21.png)
![image](https://user-images.githubusercontent.com/70942617/187920828-43e867db-a158-4f0b-961b-63e964973eb9.png)

New:
![image](https://user-images.githubusercontent.com/4244591/188103536-ec780171-61d0-4925-9271-89dd6b9bd55f.png)
![image](https://user-images.githubusercontent.com/4244591/188103428-c0ba693a-c002-4118-be09-df9026440a44.png)
![image](https://user-images.githubusercontent.com/4244591/188103441-67fa3805-5768-44c8-bce4-0983febc673c.png)
![image](https://user-images.githubusercontent.com/4244591/188103452-39095e4c-169c-46f8-a88b-3ba0d49a273f.png)
![image](https://user-images.githubusercontent.com/4244591/188103463-2ee1b6da-25c6-4f2e-bfaf-07afa6c9f4de.png)
![image](https://user-images.githubusercontent.com/4244591/188103474-26797c93-3b34-442f-bf5f-059114be2465.png)

aMannus committed:
Columns in childs & code style changes

aMannus committed:
Incompatible checkboxes

aMannus committed:
Code styling

leggettc18 commented:
So while I do like the per-column scrollbars in general, I do wonder if it wouldn't be better to have the "Main Rules" tab changed to a "World Settings" tab, and have the Open Settings, World Settings (i.e. Entrance Rando, Bombchu Drops, etc.), and Item Pool & Hint stuff there, and make "Other" be a "Shuffle Settings" tab, move "Shuffle Settings" and "Shuffle Dungeon Items" there, and break up the current "Shuffle Settings" tab into two columns. One for the big World Shuffle settings (Tokens, Songs, Scrubs, Shops, and later Pots, basically anything that has a drop down), one for the smaller settings (basically the ones that are just a Checkbox now, plus future magic beans, merchant shuffle, etc.) and then one for the Dungeon Items, not necessarily in that order. I'm not 100% sold on that but I do see the "Shuffle Settings" column as it is now getting huge in a few more updates. Magic Beans is PR'd, Shopsanity is on the way, Merchant Shuffle will likely be pretty soon, etc.

Also this is a minor detail, but regardless of what else happens the separator between Tokensanity and Nighttime GS Expect Sun's Song should probably not be there, those settings should definitely be more visually linked imo.

aMannus commented:
So far I've been sort of trying to follow N64's menu, but I agree it might be time to look into a completely new layout instead.

That said, with your suggestion, we'd have timesavers leftover with no really good place to put those, and having a new tab for just those is kind of overkill.

I'll think about it some more as well and see if I can come up with something.

briaguya0 commented:
if we're talking about rearranging everything...

# Current
## Main Rules
### Open Settings
* Forest
* Kak
* DoT
* ZF
* GF
* Bridge
* Skip Ganon's Trials
### Shuffle Settings
* Songs
* Scrub Shuffle
* Tokensanity
* Nighttime GS Sun's Song
* Cows
* Adult Trade
* Kokiri Sword
* Ocarina
* Weird Egg
* Gerudo Card
* Frogs
### Suffle Dungeon Items
* Dungeon Rewards
* Maps/Compasses
* Small Keys
* GF Keys
* Boss Keys
* Ganon's Boss Key
## Other
### Timesavers
* Cuccos to return
* Big poe count
* Skip child stealth
* Skip child zelda
* Skip epona race
* Skip tower escape
* Complete mask quest
* Enable glitch-useful cutscenes
### World Settings
* Coming soon
### Item Pool & Hint Settings
* Item pool
* Ice Traps
* Gossip Stone Hints
  * Clarity
  * Distribution
## Locations
* Not gonna list all those here
## Tricks/Glitches
* Logic Rules
* list of all coming soon
## Starting Inventory
### Starting Equipment
* Kokiri Sword
* Deku Shield
### Starting Items
* Fairy Ocarina
* Consumables
* Full Wallets
### Starting Songs
* Coming soon

**Starting Inventory**: only things that come to mind now are we should support all the starting inventory in a way similar to the locations (maybe do equipment/items etc. as tree nodes), and then move consumables/full wallet somewhere?

**Tricks and glitches**: this is all todo, i think we have a good plan or it

**Locations**: i'm happy with it for now

**Other**: timesavers column here feels kinda kitchen sinky but i guess they're all timesavers, not sure what kind of cleanup makes sense for it

**Main Rules**: not the biggest fan of rainbow bridge and skip ganon's trials in open settings, i'm more ok with gerudo fortress in there but it still doesn't make the most sense. i'm also not the biggest fan of splitting up shuffle based on dungeon items/non-dungeon items. i feel like it might make more sense to group the "sanities" together and have some of the smaller things ("shuffle kokiri sword" etc.) together

this is getting pretty big by now though, and this discussion could probably go on forever.

# as for this PR: i like that the scrollbars per column are working!

aMannus commented:
Yeah tricks/glitches and starting inventory should be handled like the locations tab, I'm not too worried about those. Consumables/wallet should move out of starting inventory, I agree.

As for this PR/discussion becoming too big, I'm fine with that. Likely I'll push out a proposal based on what you guys have said and what I think, then we can go from there. So any discussion until then is welcome. I'd rather create merge conflicts with other rando-next features once instead of multiple times ;)

So currently, based on what's been said, I think we'll need to seperate the current Main Logic tab and the Other tab into 3 tabs instead. How those tabs and the columns in them should be named I'm not sure about yet.

briaguya0 commented:
also, i'll take this opportunity to mention that i dislike the "enable randomizer" checkbox living in the randomizer settings menu. I know @PurpleHato didn't like the idea of it in the dropdown because it could lead to some strange UX with needing to go back and forth between the dropdown and the settings menu, but i think the current UX of needing to open the settings menu to turn off randomizer is pretty bad

maybe having it live in 2 places makes sense?

leggettc18 commented:
I like people needing to open up the full window in order to _enable_ randomizer, so they have the opportunity to see the Generate Button and all the settings, but I agree that having to open the window to disable the randomizer doesn't feel great. Maybe we could have a "Disable Randomizer" button in the drop down that only appears when the randomizer is enabled? Maybe we could even take it a step farther, remove the Enable Randomizer Checkbox and just enable rando automatically when generating a seed? Not sure how I feel about that pattern though.

briaguya0 commented:
> I like people needing to open up the full window in order to *enable* randomizer

i agree, not having a seed but having it enabled would lead to confusion

> just enable rando automatically when generating a seed? Not sure how I feel about that pattern though.

not the biggest fan of that, what happens if someone starts a seed, then plays vanilla for a bit, then comes back to the rando seed (or wants to make a new save with the same seed)?

i think the biggest issue is that it's not actually enabling rando, it's making it so the next save file you start will be rando'd. i'm not sure the best way to communicate that though...

leggettc18 commented:
Honestly this might be one of those rare cases where if we could somehow add an extra step to the in-game UI it would be the most clear. Imagine this, you create a new save file, and the game asks if you want it to be a Randomized or Vanilla save file. Then there would be no need for an "Enable Randomizer" checkbox at all. Rando is just always on and Users don't have to ask how to disable rando because the game asks for them. No idea if that's at all possible though just spitballing.

aMannus commented:
> Honestly this might be one of those rare cases where if we could somehow add an extra step to the in-game UI it would be the most clear. Imagine this, you create a new save file, and the game asks if you want it to be a Randomized or Vanilla save file. Then there would be no need for an "Enable Randomizer" checkbox at all. Rando is just always on and Users don't have to ask how to disable rando because the game asks for them. No idea if that's at all possible though just spitballing.

I've been thinking this as well. Ask the player to create a rando file or not whenever a seed is already generated or a previous one is found when creating a new save file. Basically all the steps it's looking for right now minus the "Enable Randomizer" CVar. As you said though, no clue how possible this is at all right now.

leggettc18 commented:
In theory we wouldn't even have to wait for the user to create a seed first. If they try to make a randomized save file without having generated a seed, I can't think of any reason we couldn't spawn the ImGui window right then and have them generate one, just don't let the file select screen move forward until a seed has been generated. Idk maybe that's too much.

aMannus commented:
That would introduce quite a few UX challenges. What do we do when a player closes the randomizer settings menu at that point?  What do we do when they do generate a seed, automatically close the settings menu? That doesn't feel right either, because maybe they generated it but wanted to look over the settings one more time. 

I think at that point it would be better to just not show the choice until a seed is generated. When people want to play rando, they'll find the rando settings menu, I'm not worried about that step.

aMannus commented:
I feel at this point specifically this part of the discussion should probably come later (or move this discussion to https://github.com/HarbourMasters/Shipwright/issues/719). I'm still not really satisfied with any of the other options we've discussed for the "Enable Randomizer" part, and the ingame solution has too many unknown variables to it right now to really decide wether it's the best way to move forward or not.

Let's focus on the layout of the tabs/columns and leave the "Enable Randomizer" checkbox as it is for now.

aMannus committed:
some stuff

aMannus committed:
Redone layout tabs/columns

aMannus committed:
Merge branch 'rando-next' into rando-settings-menu-even-moar-changes

aMannus committed:
Merge fix

aMannus committed:
Code Cleanup

aMannus committed:
Merge branch 'rando-next' into rando-settings-menu-even-moar-changes

aMannus committed:
Renamed tabs

briaguya-ai commented:
just gonna leave a comment with a before/after for the menu

# Before
![image](https://user-images.githubusercontent.com/70942617/187920628-6f2186b4-aae7-48d9-824b-db0e4dd4e4d9.png)

![image](https://user-images.githubusercontent.com/70942617/187920693-4771d916-b6d6-49e8-9197-9c99646b42d0.png)

![image](https://user-images.githubusercontent.com/70942617/187920741-0c561339-c8e4-4b44-a1dc-f5b190f5ecbc.png)

![image](https://user-images.githubusercontent.com/70942617/187920795-f7004386-9820-4166-8b17-06847e88fb21.png)

![image](https://user-images.githubusercontent.com/70942617/187920828-43e867db-a158-4f0b-961b-63e964973eb9.png)

# After

![image](https://user-images.githubusercontent.com/70942617/187921262-e2e44118-0a13-4e0f-8c15-96e7a76ad943.png)

![image](https://user-images.githubusercontent.com/70942617/187921305-80926322-5cc8-4efd-9b5d-f91deb2eb948.png)

![image](https://user-images.githubusercontent.com/70942617/187921344-312b0c0c-b6ff-4db9-bd8b-2569f0d499e6.png)

![image](https://user-images.githubusercontent.com/70942617/187921428-e908bd4b-054a-437d-806a-7bc089648dad.png)

![image](https://user-images.githubusercontent.com/70942617/187921460-7c9d87a5-5ddf-4e24-97c8-2ac907ffea9a.png)

![image](https://user-images.githubusercontent.com/70942617/187921495-2769f2b1-92cf-4255-a4b3-ae05893d8dff.png)

aMannus committed:
Name changes and small layout changes

aMannus committed:
Merge branch 'rando-next' into rando-settings-menu-even-moar-changes

aMannus commented:
@briaguya-ai I added all the screenshots of the old menu and updated ones for the new one to the OP, so you can remove your comment with the screenshots.

briaguya0 commented:
not sure how i feel about "Gameplay" as the tab name, do hints and item pool settings really fit that well into gameplay?

names are hard...

also not sure about "Items," maybe "Shuffle" fits better for that in my mind but I'm really not sure.

might be worth trying to get some more eyes on this one to get some suggestions for names.

leggettc18 commented:
I think Gameplay is fine. Item Pool may effect how long your seed takes, Hint Settings affect what hints you can find and how quickly you know where certain items are, and those settings also don't directly affect randomization. Items is also apt, because that tab applies to where you'll find the items. Shuffle isn't good because the World tab will include the options for shuffling Entrances, so it would feel weird to have them outside of the Shuffle tab if it was named that. "Item Shuffle" would work but it would stick out a bit due to the length. Maybe that's OK though.

briaguya0 commented:
i'm fine with keeping it as is and dealing with not liking some of the names, but I'd really like to hear some new ideas. At this point they all work and are fine, but i don't like them. I was really hoping someone might have some suggestions that we end up really liking.

aMannus commented:
"Gameplay" is the one I had the most doubts about, also because of the Item Pool and Hints column. I said as much in discord before, but forgot to mention it here. "Items" I'm personally quite happy with, since it really does only have options for shuffling items. I had it as "Shuffle Items" before, but I tried to trim down all the tab names because they were becoming quite big.

But yeah I'm open for ideas for possible alternatives of course.

briaguya0 commented:
I guess "Items" to me makes me want to click it and see a tab similar to "Locations" but with 3 columns with every item and "Shuffled/Vanilla/Start With"

when in reality items has stuff like tokensanity

aMannus commented:
> I guess "Items" to me makes me want to click it and see a tab similar to "Locations" but with 3 columns with every item and "Shuffled/Vanilla/Start With"
> 
> when in reality items has stuff like tokensanity

I'm not quite sure why that's your expectation with that name. Tokensanity is still shuffling skull tokens, which are items. Would "Shuffle Items" or "Item Shuffle" change that expectation for you?

briaguya0 commented:
> > I guess "Items" to me makes me want to click it and see a tab similar to "Locations" but with 3 columns with every item and "Shuffled/Vanilla/Start With"
> > when in reality items has stuff like tokensanity
> 
> I'm not quite sure why that's your expectation with that name. Tokensanity is still shuffling skull tokens, which are items. Would "Shuffle Items" or "Item Shuffle" change that expectation for you?

not really, i guess i feel like it being called items makes me thing about the *items*, where as stuff like tokensanity makes me think about the *checks*

maybe just calling the tab "Checks"? idk

leggettc18 commented:
I personally didn't like calling the tab "Shuffle" because of entrance shuffle being in a different tab, but looking at n64rando, they have the equivalent settings in a column labeled "Shuffle" and have entrance shuffle stuff outside of that in their "World" column. I personally still don't like "Shuffle" but there is a precedent for it.

aMannus commented:
"Checks" is terminology people that aren't familiar with rando probably don't understand. I wanted to use that word in some tooltips a while ago as well but refrained from doing so because of that.

I've been trying to think of other names but so far, I'm drawing a complete blank.

"Shuffle" clashes with shuffle entrances. "Item Shuffle" or "Shuffle Items" doesn't solve the problem. "Checks" is terminology. Calling it "Main Rules" or something along those lines like N64 doesn't fit. I'm kind of out of ideas at this point.

briaguya0 commented:
we could discuss names forever, little PRs to change names will be fast if we think of better ones. calling it good to go
